enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,634 @@
+    termination: edge
+  to:
+    kind: Service
+    name: hello-openshift
+  httpHeaders:
+     action: Set
+     set:
+       response:
+       - name: X-Frame-Options
+         value: DENY
+         action: Replace
+       - name: X-XSS-Protection
+         value: 1;mode=block
+         action: Replace
+```
+5. Cluster admin wants to delete HTTP response headers at global level which need to be sent to all clients.
+6. Cluster admin edits/creates the CR of the Ingress Controller having Spec as follows -
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: default
+  namespace: openshift-ingress-operator
+spec:
+   httpHeaders:
+     action: Delete
+     delete:  
+       response:
+       - name: X-Frame-Options
+       - name: X-XSS-Protection
+```
+7. Users/Devops/Developers want to delete HTTP response header which need to be sent only to the client hitting
+   the route with the HTTP Response Header values.
+8. Users/Devops/Developers edits/creates the Route having Spec as follows -
+```yaml
+apiVersion: route.openshift.io/v1
+kind: Route
+metadata:
+  labels:
+    type: sharded 
+  name: hello-openshift-edge
+  namespace: hello-openshift
+spec:
+  subdomain: hello-openshift 
+  tls:
+    termination: edge
+  to:
+    kind: Service
+    name: hello-openshift
+  httpHeaders:
+    action: Delete
+    delete:
+      response:
+      - name: X-Frame-Options
+      - name: X-XSS-Protection
+```
+#### Variation [optional]
+
+### API Extensions
+
+A) Set(Replace)/Append/Delete HTTP headers in the global section of haproxy.config via IngressController's Spec.
+The existing Ingress Controller API `HTTPHeaders *IngressControllerHTTPHeaders` is extended by adding an field for action to be performed,
+and field for headers to be set/replaced/appended/deleted.
+
+```go
+// IngressControllerSpec is the specification of the desired behavior of the
+// IngressController.
+type IngressControllerSpec struct {
+    // httpHeaders defines policy for HTTP headers.
+    //
+    // If this field is empty, the default values are used.
+    //
+    // +optional
+    HTTPHeaders *IngressControllerHTTPHeaders `json:"httpHeaders,omitempty"`
+}
+```
+
+```go
+
+type IngressControllerHTTPHeaders struct {
+	
+	// +unionDiscriminator
+	// +kubebuilder:validation:Enum:=Set;Delete
+	// +kubebuilder:validation:Required
+	Action IngressControllerSetHTTPHeaderAction `json:"action,omitempty"`
+
+	// Set defines HTTP headers that should be set or replaced or appended.
+	// If this field is empty, no headers are added.
+	//
+	// Note that this option only applies to cleartext HTTP connections
+	// and to secure HTTP connections for which the ingress controller
+	// terminates encryption (that is, edge-terminated or reencrypt
+	// connections).  Headers cannot be set(replace)/append for TLS passthrough
+	// connections.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	Set IngressControllerSetHTTPHeaders `json:"set,omitempty"`
```

Comment:
Note that all the union members [must be pointers](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#writing-a-union-in-go).

Reply:
changed